spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/core/cloudfoundry/GcpCloudFoundryEnvironmentPostProcessor.java

```diff
@@ -0,0 +1,129 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.autoconfigure.core.cloudfoundry;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+import com.google.common.collect.ImmutableMap;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.cloud.CloudPlatform;
+import org.springframework.boot.context.config.ConfigFileApplicationListener;
+import org.springframework.boot.env.EnvironmentPostProcessor;
+import org.springframework.boot.json.JsonParser;
+import org.springframework.boot.json.JsonParserFactory;
+import org.springframework.core.Ordered;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.core.env.PropertiesPropertySource;
+
+/**
+ * @author João André Martins
+ */
+public class GcpCloudFoundryEnvironmentPostProcessor
+		implements EnvironmentPostProcessor, Ordered {
+
+	public static final String VCAP_SERVICES_ENVVAR = "VCAP_SERVICES";
+
+	private static final Log LOGGER =
+			LogFactory.getLog(GcpCloudFoundryEnvironmentPostProcessor.class);
+
+	private static final String SPRING_CLOUD_GCP_PROPERTY_PREFIX = "spring.cloud.gcp.";
+
+	private JsonParser parser = JsonParserFactory.getJsonParser();
+
+	private int order = ConfigFileApplicationListener.DEFAULT_ORDER - 1;
+
+	private enum GcpCfService {
+
+		pubsub("google-pubsub",
+				ImmutableMap.of("ProjectId", "project-id",
+				"PrivateKeyData", "credentials.encodedKey",
+				"topic_name", "topic-name",
+				"subscription_name", "subscription-name")),
+		storage("google-storage",
+				ImmutableMap.of("ProjectId", "project-id",
+				"PrivateKeyData", "credentials.encodedKey",
+				"bucket_name", "bucket-name"));
+
+		private String cfServiceName;
+		private Map<String, String> cfPropNameToGcp;
+
+		GcpCfService(String cfServiceName,
+				Map<String, String> cfPropNameToGcp) {
+			this.cfServiceName = cfServiceName;
+			this.cfPropNameToGcp = cfPropNameToGcp;
+		}
+
+		public String getCfServiceName() {
+			return this.cfServiceName;
+		}
+
+		public Map<String, String> getCfPropNameToGcp() {
+			return this.cfPropNameToGcp;
+		}
+	}
+
+	@Override
+	public void postProcessEnvironment(ConfigurableEnvironment environment,
+			SpringApplication application) {
+		if (CloudPlatform.CLOUD_FOUNDRY.isActive(environment)) {
+			Map<String, Object> vcapMap =
+					this.parser.parseMap(environment.getProperty(VCAP_SERVICES_ENVVAR));
+
+			Properties gcpCfPubSubProps = new Properties();
+
+			Arrays.stream(GcpCfService.values()).forEach(cfService ->
+					gcpCfPubSubProps.putAll(retrieveCfProperties(vcapMap, cfService)));
+
+			environment.getPropertySources()
+					.addLast(new PropertiesPropertySource("gcpCf", gcpCfPubSubProps));
```

Comment:
I think that's wrong. You want that the cloud information overrides anything you've set in your own config right? That's what's happening for the  `DataSource` support in CF (your `application.properties` refer to some DB you use locally but you want to use the one provided by the cloud when you push your app there.

Reply:
Ack. Should we take the same care to not override the command line parameters, like in https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/cloud/CloudFoundryVcapEnvironmentPostProcessor.java#L126 ?